src/interp/interp-wasi.cc

```diff
@@ -0,0 +1,250 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/interp/interp-wasi.h"
+#include "src/interp/interp-util.h"
+
+#ifdef WITH_WASI
+
+#include "uvwasi.h"
+
+#include <cinttypes>
+#include <unordered_map>
+
+using namespace wabt;
+using namespace wabt::interp;
+
+namespace {
+
+// Types that align with WASI specis on size and alignment.
+// TODO(sbc): Auto-generate this from witx.
+typedef uint32_t __wasi_size_t;
+typedef uint32_t __wasi_ptr_t;
+
+_Static_assert(sizeof(__wasi_size_t) == 4, "witx calculated size");
+_Static_assert(_Alignof(__wasi_size_t) == 4, "witx calculated align");
+_Static_assert(sizeof(__wasi_ptr_t) == 4, "witx calculated size");
+_Static_assert(_Alignof(__wasi_ptr_t) == 4, "witx calculated align");
+
+struct __wasi_iovec_t {
+  __wasi_ptr_t buf;
+  __wasi_size_t buf_len;
+};
+
+_Static_assert(sizeof(__wasi_iovec_t) == 8, "witx calculated size");
+_Static_assert(_Alignof(__wasi_iovec_t) == 4, "witx calculated align");
+_Static_assert(offsetof(__wasi_iovec_t, buf) == 0, "witx calculated offset");
+_Static_assert(offsetof(__wasi_iovec_t, buf_len) == 4, "witx calculated offset");
+
+class WasiInstance {
+ public:
+  WasiInstance(Instance::Ptr instance,
+               uvwasi_s* uvwasi,
+               Memory* memory,
+               Stream* trace_stream)
+      : trace_stream(trace_stream),
+        instance(instance),
+        uvwasi(uvwasi),
+        memory(memory) {}
+
+  Result proc_exit(const Values& params, Values& results, Trap::Ptr* trap) {
+    const Value arg0 = params[0];
+    uvwasi_proc_exit(uvwasi, arg0.i32_);
+    return Result::Ok;
+  }
+
+  Result fd_write(const Values& params, Values& results, Trap::Ptr* trap) {
+    const Value arg0 = params[0];
+    int32_t iovptr = params[1].i32_;
+    int32_t iovcnt = params[2].i32_;
+    __wasi_iovec_t* wasm_iovs = getMemPtr<__wasi_iovec_t>(iovptr, iovcnt, trap);
+    if (!wasm_iovs) {
+      return Result::Error;
+    }
+    uvwasi_ciovec_t* iovs = new uvwasi_ciovec_t[iovcnt];
+    for (int i = 0; i < iovcnt; i++) {
+      iovs[0].buf_len = wasm_iovs[0].buf_len;
+      iovs[0].buf =
+          getMemPtr<uint8_t>(wasm_iovs[0].buf, wasm_iovs[0].buf_len, trap);
+      if (!iovs[0].buf) {
+        return Result::Error;
+      }
+    }
+    __wasi_ptr_t* out_addr = getMemPtr<__wasi_ptr_t>(params[3].i32_, 1, trap);
+    if (!out_addr) {
+      return Result::Error;
+    }
+    uvwasi_fd_write(uvwasi, arg0.i32_, iovs, iovcnt, out_addr);
+    delete[] iovs;
+    return Result::Ok;
+  }
+
+  // The trace stream accosiated with the instance.
+  Stream* trace_stream;
+
+  Instance::Ptr instance;
+ private:
+  template <typename T>
+  T* getMemPtr(uint32_t address, uint32_t size, Trap::Ptr* trap) {
```

Comment:
Hm, we may want to add this to the interp API instead, now that I think of it. But it's OK to have here for now.